scripts/compare_license.py

```diff
@@ -0,0 +1,166 @@
+    Exluding .po files as they create only false positives
+    """
+    stdout = os.popen('rg --files-without-match "@author Xournal++" `rg -i -e "copyright" -e "license" -l` | rg -v "\.po"').readlines()
+    files = [f.strip() for f in stdout]
+    return set(files)
+
+def get_changed_files_since(git_hash:str):
+    stdout = os.popen(f'git diff {git_hash} HEAD --name-only').readlines()
+    files = [f.strip() for f in stdout]
+    return set(files)
+
+# I: Add an entry if a file is detected automatically as a file with special
+# license/copyright, but which is actually licensed/copyrighted under the same
+# license/copyright as xournalpp.
+# Please add a short comment explaining why it's whitelisted
+def get_whitelist_not_listed():
+    """Whitelist for files containing the searched for substrings but
+    are not necessary for the copyright.txt"""
+    white_list = set()
+    white_list.add("ABOUT-NLS") # false positive
+    white_list.add("copyright.txt") # copyright/license summary file
+    white_list.add("CMakeLists.txt") # false positive
+    white_list.add("LICENSE") # main license file
+    white_list.add("rpm/fedora/xournalpp.spec") # false positive
+    white_list.add("windows-setup/xournalpp.nsi") # false positive
+    white_list.add("ui/about.glade") # false positive
+    white_list.add("src/win32/xpp.rc.in") # false positive
+    return white_list
+
+# II: Add an entry to the whitelist if you added a file which has special
+# licensing/copyright but does not contain any of the substrings used to 
+# automatically identify such files
+# The rational should be explained in the copyright.txt file itself.
+# Do not use comments in this file to explain the rational.
+def get_whitelist_not_found():
+    """Whitelist for files listed in copyright.txt but do not include 
+    the searched for substrings"""
+    white_list = set()
+    white_list.add("*")
+    white_list.add("debian/changelog")
+    white_list.add("debian/compat")
+    white_list.add("debian/control")
+    white_list.add("debian/docs")
+    white_list.add("debian/package_description")
+    white_list.add("debian/rules")
+    white_list.add("debian/source/format")
+    white_list.add("ui/pixmaps/application-x-xojpp.svg")
+    white_list.add("ui/pixmaps/application-x-xopp.svg")
+    white_list.add("ui/pixmaps/application-x-xopt.svg")
+    white_list.add("ui/pixmaps/com.github.xournalpp.xournalpp.png")
+    white_list.add("ui/pixmaps/com.github.xournalpp.xournalpp.svg")
+    white_list.add("ui/pixmaps/gnome-mime-application-x-xopp.svg")
+    white_list.add("ui/pixmaps/gnome-mime-application-x-xopt.svg")
+    white_list.add("ui/pixmaps/xopt.svg")
+    white_list.add("ui/iconsColor-dark/*")
+    white_list.add("ui/iconsColor-light/*")
+    return white_list
+
+
+# III: Update git commit hash to current commit once you checked
+# that the changes do not affect the licensing information in copyright.txt
+last_checked_git_commit_hash = "99df96847cafb7510d4e302acda0007bce009eb7"
```

Comment:
Could you modify this line to automatically get the commit hash of the last official release? This would simplify usage by quite a bit.

Reply:
In theory yes but, then the script would always exit with `1` if any file changed since last release, and there's no clear way to set the status to "rechecked, all good".

The workflow that is intended with the script is rather.

 - Run before next release
 - if it exits with 0: all good
 - if it exits with 1:
     - check the files that have been changed
     - update licenses if required
     - update `last_checked_git_commit_hash` to last git commit hash on master
     - merge change to master

This way when you release this script should always yield a `0` exit code, if not one shouldn't yet release.

Does that make sense, or would you have a different idea for the workflow?

Reply:
Could we instead create a script that can be automatically run on a merge request and check if a license changed in a file or one was added without license? That would make it obvious to the person providing the merge request to add this info.

Reply:
This is actually how the script could be used as well, however there's one caveat why it might make sense to only make it mandatory to run it pre release.

Following Usecase is problematic:
 1. You updated a file with license
 2. You run script on PR -> script raises warning
 3. You recheck and update the commit hash in the script to the last one of your PR
 4. You commit this change to your PR -> Now the PR builds without problems
 5. You squash merge 
 6. :boom: 

In this case the `last_checked_git_commit_hash` would no longer be valid and the script would fail.

I see two solutions for this problem:
 1. Do never squash PRs (probably not that great in the long run)
 2. Allow to merge PRs even if the script fails (it would then be expected/ideal if the respective developer would raise a new PR updating the `last_checked_git_commit_hash` based on his merged last PR)

For me Nr. 2 seems like the best way. However, I'm not sure if this is an easy enough workflow to expect that every contributor knows about this.

Hence, I would say for release builds the script should never fail (as the main contributors probably know how to fix it). 
For usual PRs it's just a warning but doesn't prevent a merge. In the worst case one would need to ask the person who merged without updating the `last_checked_git_commit_hash` for details prior to the release if something is unclear.

Reply:
IFor PRs the commit hash should not be updated manually. This would equal a merge conflict for every PR we have.

Instead use HEAD of master as base commit and check for all files that are added/removed/changed whether the file has a license and the copyright data is accurate. We could enforce SPDX identifiers for this in all new/modified files.

Reply:
For now this script is absolutely okay. Delaying the release it not worth this commit. My point is more for the long run.

Main goal for the future should be that no code gets submitted without correct copyright notice. Otherwise code from this specific commit is not covered by the copyright (from a very strict point of view). I know of projects that rebased the whole project because of this.

I'd recommend SPDX tags for concise and consistent copyright notices throughout all files. They are currently being established as kind of a standard throughout open source projects as more and more teams include them in their sources.